### PR TITLE
Create a new GoToTypeDefinition endpoint

### PR DESCRIPTION
**Pull request by jtschuster**
Creates a new endpoint to go the definition of the type of a local variable, parameter, field, or property. The implementation borrows heavily from the GotoDefinition endpoint. Fixes issue https://github.com/OmniSharp/omnisharp-roslyn/issues/2297

This endpoint should be helpful in C# workflow, especially when using implicitly typed variables that don't have the name of the type anywhere in the file. This also provides an endpoint to use with the VS Code command `editor.action.goToTypeDefinition`. 

Example:
file1.cs:
```C#
class C {
    C NamedConstructor() { return new C(); }
}
```
file2.cs:
```C#
class D {
    void method() 
    {
        var variableName = NamedConstructor();
    }
}
```
GoToTypeDefinition at the position of `variableName` would return the definition of the class `C` in file1.cs.

### CODE REVIEW

**Comment by jtschuster** [comment]
@filipw @JoeRobich You seem to be the most active in this repository, could you take a look or let me know if there's anything I need to do for a PR?

**Comment by JoeRobich** [comment]
> @filipw @JoeRobich You seem to be the most active in this repository, could you take a look or let me know if there's anything I need to do for a PR?

I will take a look at this next week.

**Comment by jtschuster** [comment]
@JoeRobich Just checking back in. No rush, just don't want this to get lost.

File: src/OmniSharp.Abstractions/OmniSharpEndpoints.cs [hunk 1]
```diff
@@ -77,6 +77,8 @@ public static class V2
             public const string Highlight = "/v2/highlight";
 
             public const string GotoDefinition = "/v2/gotodefinition";
+
+            public const string GotoTypeDefinition = "/v2/gototypedefinition";
```
Comment (JoeRobich) [inline src/OmniSharp.Abstractions/OmniSharpEndpoints.cs:81]: @filipw Since this is the first iteration of this endpoint, is it ok to be under V2?

**Review by JoeRobich — APPROVED** [review]
Thanks for the contribution!

File: src/OmniSharp.Abstractions/OmniSharpEndpoints.cs (hunk 1 above)
Comment (filipw) [reply src/OmniSharp.Abstractions/OmniSharpEndpoints.cs:81]: I think it should just be `/gototypedefinition`, we do not version at API level but on endpoint level.
for example `/completion` is the modern version of intellisense and it was not versioned as `v2`

File: src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoTypeDefinitionHandler.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+{
+    [OmniSharpHandler(OmniSharpEndpoints.V2.GotoTypeDefinition, Constants.LanguageNames.Cake), Shared]
+    public class GotoTypeDefinitionV2Handler : CakeRequestHandler<GotoTypeDefinitionRequest, GotoTypeDefinitionResponse>
+    {
+        private readonly MetadataExternalSourceService _metadataExternalSourceService;
+
+        [ImportingConstructor]
+        public GotoTypeDefinitionV2Handler(
+            OmniSharpWorkspace workspace,
+            MetadataExternalSourceService metadataExternalSourceService)
```
[the remark this answers is not in the document]
Comment (JoeRobich) [reply src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoTypeDefinitionHandler.cs:27]: This PR seems to be mostly a cut and paste of the [GotoDefinitionV2Handler](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoDefinitionV2Handler.cs). Perhaps we can update all the Navigation handlers as a follow up.

File: src/OmniSharp.Abstractions/OmniSharpEndpoints.cs (hunk 1 above)
Comment (jtschuster) [reply src/OmniSharp.Abstractions/OmniSharpEndpoints.cs]: Okay, it's updated accordingly

File: src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoTypeDefinitionHandler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtschuster) [reply src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoTypeDefinitionHandler.cs]: > This file seems to be mostly a cut and paste of the [GotoDefinitionV2Handler](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoDefinitionV2Handler.cs).
Yeah, it was a copy and paste. I've updated to use the `ExternalSourceServiceFactory`.

File: src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoTypeDefinitionHandler.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+                    continue;
+                }
+
+                if (!request.WantMetadata)
+                {
+                    continue;
+                }
+
+                var aliasLocations = await GotoTypeDefinitionHandlerHelper.GetAliasFromMetadataAsync(
```
[the remark this answers is not in the document]
Comment (jtschuster) [reply src/OmniSharp.Cake/Services/RequestHandlers/Navigation/GotoTypeDefinitionHandler.cs]: Thanks for the tip, I've updated to use the `IExternalSourceService` from the `ExternalSourceServiceFactory`.

**Review by JoeRobich — APPROVED** [review]
Thanks @jtschuster!

**Review by filipw — APPROVED** [review]

**Comment by filipw** [comment]
thanks!

**Pull request merged**